### PR TITLE
feat(api): add crud hatchery and rbac (hatchery + region)

### PR DESCRIPTION
**Pull request by sguiheux**
1. Description
1. Related issues
1. About tests
1. Mentions

@ovh/cds

### CODE REVIEW

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#18113.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test UI ✔

File: docs/content/hosting/upgrade/migrate_0.53.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+* Migrate all existing users in the organization
+```shell
+cdsctl admin organization user-migrate my-organization
+```
+
+
+## Upgrading to 0.53.0
+
+As we are updating DB table around authentication consumer, you will need to completly shutdown your CDS instances and all µservices.
```
Comment (yesnault) [inline docs/content/hosting/upgrade/migrate_0.53.md:76]:
```suggestion
As we are updating DB table around authentication consumer, you will need to completely shutdown your CDS instances and all µservices.
```

File: docs/content/hosting/upgrade/migrate_0.53.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+* Shutdown CDS and all µservices
+* Run the database migration, documentation on https://ovh.github.io/cds/hosting/database/
+* Start 1 (scale to 1 if you usually use multiple instances) CDS API, check if there is no error on migration, with cdsctl admin migration list
+  * There are two migrations to check: 'OrganizationMigration' and 'ConsumerMigration'
+  * Migration can take a few minutes(between 1 and 5) depending on the number of users you have.
+* Scale up CDS API if you usually use multiple instances 
+* Start the others µservices
```
Comment (yesnault) [inline docs/content/hosting/upgrade/migrate_0.53.md:84]:
```suggestion
* Start other µservices 
```

File: docs/content/hosting/upgrade/migrate_0.53.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+## Upgrading to 0.53.0
+
+As we are updating DB table around authentication consumer, you will need to completly shutdown your CDS instances and all µservices.
+
+* Shutdown CDS and all µservices
+* Run the database migration, documentation on https://ovh.github.io/cds/hosting/database/
+* Start 1 (scale to 1 if you usually use multiple instances) CDS API, check if there is no error on migration, with cdsctl admin migration list
+  * There are two migrations to check: 'OrganizationMigration' and 'ConsumerMigration'
+  * Migration can take a few minutes(between 1 and 5) depending on the number of users you have.
```
Comment (yesnault) [inline docs/content/hosting/upgrade/migrate_0.53.md:82]:
```suggestion
  * Migration can take a few minutes (between 1 and 5) depending on the number of users you have.
```

File: docs/content/hosting/upgrade/migrate_0.53.md [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+```
+
+
+## Upgrading to 0.53.0
+
+As we are updating DB table around authentication consumer, you will need to completly shutdown your CDS instances and all µservices.
+
+* Shutdown CDS and all µservices
+* Run the database migration, documentation on https://ovh.github.io/cds/hosting/database/
+* Start 1 (scale to 1 if you usually use multiple instances) CDS API, check if there is no error on migration, with cdsctl admin migration list
```
Comment (yesnault) [inline docs/content/hosting/upgrade/migrate_0.53.md:80]:
```suggestion
* Start 1 (scale to 1 if you usually use multiple instances) CDS API, check if there is no error on migration, with `cdsctl admin migration list`
```

**Review by yesnault — CHANGES_REQUESTED** [review]

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#18118.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘

**Review by yesnault — APPROVED** [review]

**Review by richardlt — APPROVED** [review]

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#18127.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test UI ✔

File: engine/api/rbac/rule_region.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+	if auth == nil {
+		return sdk.WithStack(sdk.ErrForbidden)
+	}
+
+	hasRole, err := HasRoleOnRegionAndUserID(ctx, db, role, auth.AuthConsumerUser.AuthentifiedUser, regionIdentifier)
+	if err != nil {
+		return err
+	}
+
+	log.RegisterField(LogFieldRole)
```
Comment (fsamin) [inline engine/api/rbac/rule_region.go:23]: try to registerField only once (registering is playing with a mutex)

File: engine/api/v2_region.go [hunk 6]
```diff
@@ -54,16 +54,37 @@ func (api *API) postRegionHandler() ([]service.RbacChecker, service.Handler) {
...
+			var err error
+
+			if consumer.Admin() {
+				trackSudo(ctx, w)
+				regions, err = region.LoadAllRegions(ctx, api.mustDB())
+				if err != nil {
+					return err
+				}
+			} else {
+				rbacRegions, err := rbac.LoadRegionIDsByRoleAndUserID(ctx, api.mustDB(), sdk.RegionRoleRead, consumer.AuthConsumerUser.AuthentifiedUserID)
```
Comment (fsamin) [inline engine/api/v2_region.go:69]: "Read a region" is a role ?

File: sdk/rbac.go [hunk 7]
```diff
@@ -15,15 +15,25 @@ const (
 	// Project Role
 	ProjectRoleRead   = "read"
 	ProjectRoleManage = "manage"
+
+	// Hatchery Role
+	HatcheryRoleSpawn = "spawn"
```
Comment (fsamin) [inline sdk/rbac.go:20]: start worker ?

File: sdk/rbac.go [hunk 8]
```diff
@@ -15,15 +15,25 @@ const (
...
+
+	// Region Role
+	RegionRoleRead   = "read"
```
Comment (fsamin) [inline sdk/rbac.go:23]: what are we able to do when we "read" a region ?

File: sdk/rbac.go [hunk 9]
```diff
@@ -15,15 +15,25 @@ const (
...
+	RegionRoleUse    = "use"
```
Comment (fsamin) [inline sdk/rbac.go:24]: execute jobs ?

**Review by fsamin — CHANGES_REQUESTED** [review]

File: sdk/rbac.go (hunk 9 above)
Comment (sguiheux) [reply sdk/rbac.go:24]: => execute

File: engine/api/v2_region.go (hunk 6 above)
Comment (sguiheux) [reply engine/api/v2_region.go:69]: replace by list

**Review by fsamin — APPROVED** [review]

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#18145.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test UI ✔

**Pull request merged**